eth/vm/logic/flow.py

```diff
@@ -57,3 +60,39 @@ def gas(computation: BaseComputation) -> None:
     gas_remaining = computation.get_gas_remaining()
 
     computation.stack_push_int(gas_remaining)
+
+
+def beginsub(computation: BaseComputation) -> None:
+    raise OutOfGas("Error: at pc={}, op=BEGINSUB: invalid subroutine entry")
+
+
+def jumpsub(computation: BaseComputation) -> None:
+    sub_loc = computation.stack_pop1_int()
+    code_range_length = computation.code.__len__()
+
+    if sub_loc >= code_range_length:
+        raise InvalidJumpDestination(
+            "Error: at pc={}, op=JUMPSUB: invalid jump destination".format(
+                computation.code.program_counter)
+        )
+
+    if computation.code.is_valid_opcode(sub_loc):
```

Comment:
~Hm, this valid opcode check looks wrong to me.~ Ah, interesting, this was not made clear by the spec, IMO. But discussion seems to support the idea that the validity check is required: https://ethereum-magicians.org/t/eip-2315-simple-subroutines-for-the-evm/3941/70?u=carver

So... maybe the only thing to do here is to add an explicit test for `JUMPSUB`ing into tho data section of a `PUSH`.